src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl_ModelComputation.cs

```diff
@@ -0,0 +1,257 @@
+
+            var filePath = GetFilePath();
+            if (filePath is null)
+                return null;
+
+            // Ensure we switch to the threadpool before calling ComputeUIModelAsync. It ensures
+            // that fetching and processing the document model is not done on the UI thread.
+            await TaskScheduler.Default;
+
+            var model = await ComputeUIModelAsync().ConfigureAwait(false);
+
+            if (model is not null)
+                StartUpdateUIModelTask();
+
+            return model;
+
+            async Task<DocumentSymbolModel?> ComputeUIModelAsync()
+            {
+                var response = await DocumentOutlineHelper.DocumentSymbolsRequestAsync(
+                    textBuffer, LanguageServiceBroker, filePath, cancellationToken).ConfigureAwait(false);
+
+                if (response is null)
+                    return null;
+
+                var responseBody = response.ToObject<DocumentSymbol[]>();
+                if (responseBody is null)
+                    return null;
+
+                var documentSymbols = DocumentOutlineHelper.GetNestedDocumentSymbols(responseBody);
+                var documentSymbolItems = DocumentOutlineHelper.GetDocumentSymbolItems(documentSymbols);
+                return new DocumentSymbolModel(documentSymbolItems, originalSnapshot);
+            }
+
+            string? GetFilePath()
+            {
+                ThreadingContext.ThrowIfNotOnUIThread();
+                if (EditorAdaptersFactoryService.GetBufferAdapter(textBuffer) is IPersistFileFormat persistFileFormat &&
+                    ErrorHandler.Succeeded(persistFileFormat.GetCurFile(out var filePath, out _)))
+                {
+                    return filePath;
+                }
+
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Starts a new task to update the UI model.
+        /// </summary>
+        private void StartUpdateUIModelTask()
+        {
+            // 'true' value is unused.  this just signals to the queue that we have work to do.
+            _updateUIModelQueue.AddWork(true);
+        }
+
+        /// <summary>
+        /// Filters and sorts the UI model.
+        /// </summary>
+        private async ValueTask<DocumentSymbolModel?> UpdateUIAsync(ImmutableSegmentedList<bool> unused, CancellationToken cancellationToken)
+        {
+            var model = await _computeUIModelQueue.WaitUntilCurrentBatchCompletesAsync().ConfigureAwait(false);
+            if (model is null)
+                return null;
+
+            // Switch to the UI thread to get the current search query and latest active text view.
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var searchQuery = searchBox.Text;
+
+            var activeTextView = GetLastActiveIWpfTextView();
+            if (activeTextView is null)
+                return null;
```

Comment:
when does this happen? (also, consider doc'ing).

Reply:
nit: i don't see what activeTextView is actually even used for here.  Do we need this logic?